jupyter_client/launcher.py

```diff
@@ -50,7 +55,122 @@ def launch_kernel(
     Returns
     -------
 
-    Popen instance for the kernel subprocess
+    subprocess.Popen instance for the kernel subprocess
+    """
+    proc = None
+
+    kwargs, interrupt_event = _prepare_process_args(
+        stdin=stdin, stdout=stdout, stderr=stderr, env=env, independent=independent, cwd=cwd, **kw
+    )
+
+    try:
+        # Allow to use ~/ in the command or its arguments
+        cmd = [os.path.expanduser(s) for s in cmd]
+
+        proc = Popen(cmd, **kwargs)
+    except Exception as exc:
+        _handle_subprocess_exception(exc, cmd, env, **kwargs)
+
+    _finish_process_launch(proc, stdin, interrupt_event)
+
+    return proc
+
+
+async def async_launch_kernel(
```

Comment:
Why do we need `launch_kernel` and `async_launch_kernel`?

Reply:
I wasn't sure if some provisioner authors might want sync Popen.  However, this was done prior to your refactor and I suspect we could/should just go with `async_launch_kernel` only.  It should be noted though that Windows `asyncio.create_subprocess_exec` has issues (noted in the comment) and we'd still use `Popen` under the covers for that platform.

I also noticed yesterday that [all objects from `launcher.py` are exposed](https://github.com/jupyter/jupyter_client/blob/98faaf91d00b4fbf17b2fbf0b08b1fd23adc27bf/jupyter_client/__init__.py#L10) - which implies `launch_kernel()` can be used directly - although not going through `KernelManager._launch_kernel()` doesn't make a whole lot of sense to me.  I wonder if we should just assume that's the supported use-case here and no direct use is permitted.  If so, we should probably retract `launcher.py`'s presence in `__init__.py` and can move `launcher.py` into `provisioning`.

Reply:
I think the only thing async subprocesses offer in our case is async `wait` (for the process to terminate), which I'm not sure we need since we send `SIGTERM` then `SIGKILL`, IIRC, but we never `await` forever.

Reply:
Thanks David, you are correct.  I happen to find `Popen` easier to deal with anyway and would be fine only having the original (and not offering an async version).  Using `Popen` also simplifies the WIndows case as well.  The important aspect is that the corresponding methods on `KernelProvisionerBase` remain `async`.

Reply:
But then don't we need both to be of the same type (sync or async)? It doesn't hurt to have an async `launch_kernel()` even if there is nothing async in it (but it will have to be awaited for sure).

Reply:
No.  When used, `launch_kernel()` is always embedded in a provisioner, and provisioner methods are only async.  It is no longer called directly.